lib/utils/mcputils/http.go

```diff
@@ -0,0 +1,139 @@
+/*
+ * Teleport
+ * Copyright (C) 2025  Gravitational, Inc.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package mcputils
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"mime"
+	"net/http"
+
+	"github.com/gravitational/trace"
+	"github.com/mark3labs/mcp-go/mcp"
+
+	"github.com/gravitational/teleport"
+	"github.com/gravitational/teleport/lib/utils"
+)
+
+// ServerMessageProcessor defines an interface that process JSON RPC responses
+// and notifications.
+type ServerMessageProcessor interface {
+	// ProcessResponse process a response and returns the message for client.
+	ProcessResponse(context.Context, *JSONRPCResponse) mcp.JSONRPCMessage
+	// ProcessNotification process a notification and returns the message for client.
+	ProcessNotification(context.Context, *JSONRPCNotification) mcp.JSONRPCMessage
+}
+
+// ReplaceHTTPResponse handles replacing the MCP server response for the
+// streamable HTTP transport.
+//
+// https://modelcontextprotocol.io/specification/2025-06-18/basic/transports#streamable-http
+func ReplaceHTTPResponse(ctx context.Context, resp *http.Response, processer ServerMessageProcessor) error {
+	// Nothing to replace.
+	if resp.StatusCode != http.StatusOK || resp.ContentLength == 0 {
+		return nil
+	}
+
+	mediaType, _, err := mime.ParseMediaType(resp.Header.Get("Content-Type"))
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	switch mediaType {
+	case "application/json":
+		// Single response.
+		respBody, err := utils.ReadAtMost(resp.Body, teleport.MaxHTTPRequestSize)
+		if err != nil {
+			return trace.Wrap(err)
+		}
+		respFromServer, err := unmarshalResponse(string(respBody))
+		if err != nil {
+			return trace.Wrap(err)
+		}
+		respToClient := processer.ProcessResponse(ctx, respFromServer)
+		respToClientAsBody, err := json.Marshal(respToClient)
+		if err != nil {
+			return trace.Wrap(err)
+		}
+		resp.Body = io.NopCloser(bytes.NewReader(respToClientAsBody))
```

Comment:
Does the resp.ContentLength should be also updated ?

Reply:
good catch. from i can tell from usage by reverseproxy, it shouldn't be necessary but might as well do it here.